### PR TITLE
[PEA] Opt out of non-escapable classes

### PR DESCRIPTION
**Pull request by caojoshua**
Non flow-sensative opts out of certain classes and considers them escaped. Partial escape analysis should do the sam.e

### CODE REVIEW

File: src/hotspot/share/opto/parseHelper.cpp [hunk 1]
```diff
@@ -430,7 +431,12 @@ void VirtualState::print_on(outputStream* os) const {
 }
 #endif
 
-void PEAState::add_new_allocation(Node* obj) {
+void PEAState::add_new_allocation(Node* obj, Compile *c) {
+  const TypeKlassPtr* klass_type = obj->as_Type()->type()->isa_klassptr();
```
Comment (navyxliu) [inline src/hotspot/share/opto/parseHelper.cpp]: At this point, `obj` should be CheckCastPPNode and its type is TypeInstPtr. 

In the future, we will extend add_new_allocation() for array object. Therefore, we need to assume its type is TypeOopPtr.  

I don't understand this code snippet. I understand you want to  calculate TypeKlassPtr* from obj. TypeKlasPtr is a subclass of TypePtr but not TypeOopPtr. You may take a look at the class declaration in type.hpp. It also manifests the class hierarchy.

If you want to get TypeKlassPtr*, you would get it from KlassNode. it's alloc->in(KlassNode). There's a API AllocateNode::Ideal_klass() for this purpose. 

From my reading, what you really need is `ciInstanceKlass*`
I have already get ik from TypeOopPtr below. Could you move the gating logic there?  let's focus on TypeInstance for the time being.

File: src/hotspot/share/opto/escape.hpp [hunk 2]
```diff
@@ -668,4 +668,13 @@ inline FieldNode::FieldNode(ConnectionGraph *CG, Node* n, EscapeState es, int of
   _has_unknown_base(false) {
 }
 
+inline bool typeCannotEscape(const TypeKlassPtr *ptr, const Compile *c) {
```
Comment (navyxliu) [inline src/hotspot/share/opto/escape.hpp]: I would leave EscapeAnalysis alone. I understand that you want to factor out the common logic. in my understanding, PEA has its own logic. Touching the established EA will add extra works for future reviewers.

File: src/hotspot/share/opto/parseHelper.cpp (hunk 1 above)
Comment (caojoshua) [reply src/hotspot/share/opto/parseHelper.cpp]: Yeah, I simplified it a lot. Now do the checks together with where we opt out of throwable class.

**Review by navyxliu — APPROVED** [review]
LGTM.

**Pull request merged**